### ISSUE TITLE
no-use-before-define doesn't check ArrowFunctionExpressions

### ISSUE THREAD

**Issue by btmills** [eslint/eslint#1895]
```js
// test.js
(function() {
    console.log(foo);
})();

(() => {
    console.log(foo);
})();

var foo = 42;
```

```js
// config.json
{
    "rules": {
        "no-use-before-define": 2
    },
    "ecmaFeatures": {
        "arrowFunctions": true
    }
}
```

```sh
$ eslint --reset --no-eslintrc -c config.json test.js

test.js
  3:16  error  foo was used before it was defined  no-use-before-define

✖ 1 problem (1 error, 0 warnings)
```

`foo` is used before it is defined on lines 3 and 7 of `test.js`, but only the former case is caught because `no-use-before-define` doesn't check `ArrowFunctionExpression`s.

**Comment by btmills** [comment]
Looks like this might be an escope issue?

**Comment by nzakas** [comment]
I'm not sure, it's only escope is we're using `getScope()` to figure this out.

**Comment by btmills** [comment]
Aha! I guessed that because `scope.references` was empty for arrow function expressions. We are using `getScope()`, but it turns out it's the implementation of `getScope()` that doesn't look for the arrow function expression. Working on this; PR coming soon I hope.
